projects/packages/jetpack-mu-wpcom/src/features/crawler-control/class-crawler-control.php

```diff
@@ -0,0 +1,123 @@
+<?php
+/**
+ * Crawler Control
+ *
+ * @package automattic/jetpack-mu-wpcom
+ */
+
+/**
+ * Crawler Control
+ *
+ * @package automattic/jetpack-mu-wpcom
+ */
+class Crawler_Control {
+
+	const OPTION_NAME     = 'wpcom_is_crawlable';
+	const TERMS_URL       = 'https://wp.me/Pe4R-180z';
+	const ERROR_MESSAGE   = 'Howdy! We are as excited about the treasure trove of great content on this blog as you are, and we are happy to discuss data research partnerships. Please learn more and contact us at: ' . self::TERMS_URL;
+	const X_TERMS         = 'X-Terms: Howdy! We appreciate this content too. Learn more about our data research partnerships: ' . self::TERMS_URL;
+	const BOT_USER_AGENTS = array(
+		'a8ctest',
+		'CCBot',
+		'sentibot',
+		'Google-Extended',
+		'FacebookBot',
+		'omgili',
+		'Amazonbot',
+	);
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->hooks();
+	}
+
+	/**
+	 * Hooks into WordPress.
+	 */
+	public function hooks() {
+		add_action( 'template_redirect', array( $this, 'exit_for_bots_unless_permitted' ) );
+	}
+
+	/**
+	 * Returns the user agent.
+	 *
+	 * @return string
+	 */
+	public function get_useragent() {
+		return ! empty( $_SERVER['HTTP_USER_AGENT'] ) ? sanitize_text_field( wp_unslash( $_SERVER['HTTP_USER_AGENT'] ) ) : '';
+	}
+
+	/**
+	 * Returns whether the user agent is a bot.
+	 *
+	 * @param string $user_agent The user agent.
+	 * @return bool
+	 */
+	public function is_useragent_a_bot( $user_agent ) {
+		$bots = apply_filters( 'wpcom_crawler_control_bots', self::BOT_USER_AGENTS );
+		foreach ( $bots as $bot ) {
+			if ( stristr( $user_agent, $bot ) ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Sends a header.
+	 * We have to mock it due to test env.
+	 *
+	 * @param string $header The header to send.
+	 */
+	public function header( $header ) {
+		header( $header );
+	}
+
+	/**
+	 * Handles crawler control.
+	 */
+	public function exit_for_bots_unless_permitted() {
+		if ( $this->is_crawlable() ) {
```

Comment:
Did a quick review out of curiosity, and I have a question. This seems like a more expensive check, should it be moved below the `if ( ! $user_agent )` line?

Reply:
nice callout. Turns out that this PR won't be necessary in the end though, but thanks for looking! peGwbA-1k8-p2#comment-2745